### ISSUE TITLE
A question related to IO handling of `close` in darwin

### ISSUE THREAD

**Issue by samuelgozi** [tigerbeetle/tigerbeetle#2599]
Hi, I loved the Async IO implementation, and while trying to understand the Darwin implementation, I had a question.
It seems like all ops except `timeout` and `close` use the kqueue mechanism.

When `close()` is called, in `submit()`, all calls except timeout are added to completed queue, and in `flush_io()` it seems like it loops over all the items in `self.completed`, and creates `KEvent` structs from it, except the switch over the operation_tag will throw when it encounters a close completion.

Basically what I'm saying is that it looks like there might be a bug when calling `close` in Darwin implementation of dispatcher that might cause a panic, but I'm not sure.

If not, can anyone please explain when will the logic of `do_operation` in a close event be called?
Im trying to understand if the `close` call is synchronised to happen in order with the current design.

**Comment by chaitanyabhandari** [comment]
Hey @samuelgozi, apologies for the late response! Hope the below explanation answers your questions.

`close` in the Darwin implementation wouldn't crash because of what you've described. The reason is that `flush_io` is invoked with the completions in `io_pending`, and only creates `Kevent` structs from completions in `io_pending`.
```C
    fn flush(self: *IO, wait_for_completions: bool) !void {
        var io_pending = self.io_pending.peek();
        var events: [256]posix.Kevent = undefined;

        // Check timeouts and fill events with completions in io_pending
        // (they will be submitted through kevent).
        // Timeouts are expired here and possibly pushed to the completed queue.
        const next_timeout = self.flush_timeouts();
        const change_events = self.flush_io(&events, &io_pending);
```

We only push completions for **.accept, .connect, .read, .write, .send, .recv**, and _not_ for **.close**. Therefore, it is not possible for `flush_io` to encounter a `close` completion! Check out the following code in [submit](https://github.com/tigerbeetle/tigerbeetle/blob/main/src/io/darwin.zig#L257-L269):

```C
                switch (operation_tag) {
                    .accept, .connect, .read, .write, .send, .recv => {
                        _ = result catch |err| switch (err) {
                            error.WouldBlock => {
                                _completion.next = null;
                                io.io_pending.push(_completion);
                                return;
                            },
                            else => {},
                        };
                    },
                    else => {},
                }

```
Finally, the `do_operation` for `.close` will be called at the end of `flush`, alongside the other completions.
```C
        self.completed.reset();
        while (completed.pop()) |completion| {
            (completion.callback)(self, completion);
        }
```

`completion.callback` in the aforementioned code is [onCompleteFn](https://github.com/tigerbeetle/tigerbeetle/blob/main/src/io/darwin.zig#L250-L254) in `submit`, which calls `do_operation`:
```C
        const onCompleteFn = struct {
            fn onComplete(io: *IO, _completion: *Completion) void {
                // Perform the actual operation
                const op_data = &@field(_completion.operation, @tagName(operation_tag));
                const result = OperationImpl.do_operation(op_data);

                // Requeue onto io_pending if error.WouldBlock
```
